File: TransactionRegister/Controller/CategoriesViewController.m
//
//  ViewController.m
//  TransactionRegister
//
//  Created by Eric Romrell on 9/4/16.
//  Copyright © 2016 Eric Romrell. All rights reserved.
//

#import "CategoriesViewController.h"
#import "TXTableView.h"
#import "CategoryTableViewCell.h"
#import "CategoryViewController.h"

#define MONTHS_TO_SHOW 6

@interface CategoriesViewController () <UITableViewDataSource, UITableViewDelegate>

@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;
@property (weak, nonatomic) IBOutlet TXTableView *tableView;

@property (nonatomic) NSArray<BudgetCategory *> *categories;
@property (nonatomic) NSMutableArray<NSDate *> *filterDates;
@property (nonatomic) NSDate *currentFilterDate;

@end

@implementation CategoriesViewController

-(void)viewDidLoad {
	[super viewDidLoad];
	
	NSDate *today = [NSDate date];
	NSCalendar *calendar = [NSCalendar currentCalendar];
	NSDateComponents *comp = [[NSDateComponents alloc] init];
	self.filterDates = [[NSMutableArray alloc] init];
	for (int i = 0; i <= MONTHS_TO_SHOW; i++) {
		[comp setMonth:-i];
		[self.filterDates addObject:[calendar dateByAddingComponents:comp toDate:today options:0]];
	}
	self.currentFilterDate = self.filterDates[0];
	
	[self loadDataWithDate:[NSDate date]];
}

-(void)viewWillAppear:(BOOL)animated {
	[super viewWillAppear:animated];
	
	self.tabBarController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Filter" style:UIBarButtonItemStylePlain target:self action:@selector(filter)];
	self.tabBarController.navigationItem.rightBarButtonItem = nil;
	
	[self.tableView deselectRowAtIndexPath:self.tableView.indexPathForSelectedRow animated:YES];
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
	if ([segue.identifier isEqualToString:@"category"]) {
		CategoryViewController *vc = segue.destinationViewController;
		vc.category = self.categories[[self.tableView indexPathForCell:sender].row];
	}
}

#pragma mark UITableViewDataSource/Delegate callbacks

-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
	NSDateFormatter *format = [[NSDateFormatter alloc] init];
	[format setDateFormat:@"MMM YYYY"];
	return [NSString stringWithFormat:@"Budget - %@", [format stringFromDate:self.currentFilterDate]];
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	return self.categories.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	CategoryTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
	[cell setCategory:self.categories[indexPath.row] withMainProperty:CATEGORY_NAME];
	return cell;
}

#pragma mark Custom Functions

-(void)filter {
	NSDateFormatter *format = [[NSDateFormatter alloc] init];
	[format setDateFormat:@"MMMM YYYY"];

	UIAlertController *controller = [UIAlertController alertControllerWithTitle:@"Select Month" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
	for (NSDate *date in self.filterDates) {
		[controller addAction:[UIAlertAction actionWithTitle:[format stringFromDate:date] style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
			[self loadDataWithDate:date];
		}]];
	}
	[controller addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
	[self presentViewController:controller animated:YES completion:nil];
}

-(void)loadDataWithDate:(NSDate *)date {
	[self.spinner startAnimating];
	[Client getBudgetWithDate:date callback:^(NSArray<BudgetCategory *> * _Nullable categories, TXError * _Nullable error) {
		[self.spinner stopAnimating];
		if (error) {
			[self showError:error];
		} else {
			self.currentFilterDate = date;
			self.categories = categories;
			[self.tableView reloadData];
		}
	}];
}

@end

File: TransactionRegister/Controller/CategoriesViewController.swift
//
//  NewCategoriesViewController.swift
//  TransactionRegister
//
//  Created by Eric Romrell on 1/20/17.
//  Copyright © 2017 Eric Romrell. All rights reserved.
//

import UIKit

private let DATE_FORMAT = DateFormatter(format: "MMM yyyy")

class CategoriesViewController: TXViewController, UITableViewDataSource, UITableViewDelegate {
	private static let MONTHS_TO_SHOW = 6

	@IBOutlet weak private var spinner: UIActivityIndicatorView!
	@IBOutlet weak private var tableView: UITableView!
	
	private var categories = [BudgetCategory]()
	private var filterDates = [Date]()
	private var currentFilterDate = Date()

    override func viewDidLoad() {
        super.viewDidLoad()

        let calendar = Calendar.current
		var comp = DateComponents()
		for i in 0..<CategoriesViewController.MONTHS_TO_SHOW {
			comp.month = -i
			self.filterDates.append(calendar.date(byAdding: comp, to: self.currentFilterDate)!)
		}
    }
	
	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)
		
		self.tabBarController?.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "Filter", style: .plain, target: self, action: #selector(filter))
		self.tabBarController?.navigationItem.rightBarButtonItem = nil
		
		if self.tableView.indexPathForSelectedRow != nil {
			self.tableView.deselectRow(at: self.tableView.indexPathForSelectedRow!, animated: true)
		}
		
		//Load the data, even when they are just coming back from the other tab
		loadData(date: currentFilterDate)
	}
	
	override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
		if segue.identifier == "category" {
			let vc = segue.destination as! CategoryViewController
			vc.category = self.categories[self.tableView.indexPath(for: sender as! UITableViewCell)!.row]
		}
	}
	
	func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
		return String(format: "Budget - %@", DATE_FORMAT.string(from: self.currentFilterDate))
	}
	
	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return self.categories.count
	}
	
	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell = tableView.dequeueReusableCell(withIdentifier: "cell") as! CategoryTableViewCell
		cell.setup(category: self.categories[indexPath.row], mainProperty: .CATEGORY_NAME)
		return cell
	}
	
	func filter() {
		let format = DateFormatter()
		format.dateFormat = "MMMM yyyy"
		
		let alert = UIAlertController(title: "Select Month", message: nil, preferredStyle: .actionSheet)
		for date in self.filterDates {
			alert.addAction(UIAlertAction(title: format.string(from: date), style: UIAlertActionStyle.default, handler: { (action) in
				self.loadData(date: date)
			}))
		}
		alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
		self.present(alert, animated: true, completion: nil)
	}
	
	private func loadData(date:Date) {
		self.spinner.startAnimating()
		Client.getBudget(date: date) { (categories, error) in
			self.spinner.stopAnimating()
			if error != nil {
				self.showError(error: error!)
			} else {
				self.currentFilterDate = date
				self.categories = categories!
				self.tableView.reloadData()
			}
		}
	}
}
